src/db-wrapper.ts

```diff
@@ -0,0 +1,233 @@
+    get_first_masterchain_block_id_with_seq_no_not_less_than(min_seq_no: number): Promise<string | undefined>;
+    get_masterchain_block_info_by_id(id: string): Promise<MasterChainBlockInfo | undefined>;
+    get_shardchain_block_info_by_id(id: string): Promise<ShardChainBlockInfo | undefined>;
+    get_last_masterchain_block_seq_no(): Promise<number | undefined>;
+    init_from(source: IDbWrapper, init_seq_no: number): Promise<void>;
+
+    add_BMT(bmt: BMT): Promise<void>;
+    get_BMT_by_ids(bmt_ids: BMT_IDs): Promise<BMT>;
+    remove_BMT_by_ids(bmt_ids: BMT_IDs): Promise<void>;
+}
+
+export type MasterChainBlockInfo = {
+    _key: string;
+    seq_no: number,
+    gen_utime: number,
+    prev_block_id: string,
+    shard_block_ids: string[],
+    message_ids: string[],
+    transaction_ids: string[],
+};
+
+export type ShardChainBlockInfo = {
+    _key: string,
+    gen_utime: number,
+    workchain_id: string;
+    seq_no: number,
+    shard: string,
+    prev_block_id: string,
+    prev_alt_block_id: string | null,
+    message_ids: string[],
+    transaction_ids: string[],
+};
+
+export class DbWrapper implements IDbWrapper {
+    readonly arango_db: Database;
+    readonly blocks: DocumentCollection;
+    readonly messages: DocumentCollection;
+    readonly transactions: DocumentCollection;
+    
+    constructor(config: Config) {
+        this.arango_db = new Database(config);
+        this.blocks = this.arango_db.collection("blocks");
+        this.messages = this.arango_db.collection("messages");
+        this.transactions = this.arango_db.collection("transactions");
+    }
+
+    async get_existing_message_ids_from_id_array(message_ids: string[]): Promise<string[]> {
+        const query = await this.arango_db.query(aql`
+            FOR m IN ${this.messages}
+            FILTER m._key IN ${message_ids}
+            RETURN m._key
+        `);
+
+        return await query.all();
+    }
+    
+    async get_existing_transaction_ids_from_id_array(transaction_ids: string[]): Promise<string[]> {
+        const query = await this.arango_db.query(aql`
+            FOR t IN ${this.transactions}
+            FILTER t._key IN ${transaction_ids}
+            RETURN t._key
+        `);
+
+        return await query.all();   
+    }
+
+    async get_first_masterchain_block_id_with_seq_no_not_less_than(min_seq_no: number): Promise<string | undefined> {
+        const query = await this.arango_db.query(aql`
+            FOR b IN ${this.blocks}
+            FILTER b.workchain_id == -1 && b.seq_no >= ${min_seq_no}
+            SORT b.seq_no ASC
+            LIMIT 1
+            RETURN b._key
+        `);
+
+        return await query.next();
+    }
+
+    async get_masterchain_block_info_by_id(id: string): Promise<MasterChainBlockInfo | undefined> {
+        const query = await this.arango_db.query(aql`
+            FOR b IN ${this.blocks}
+            FILTER b.workchain_id == -1 && b._key == ${id}
```

Comment:
workchain filter is redundant because of the _key is an unique value.

Reply:
If the id of shardchain block is provided, this method should throw or return undefined. Workchain filter is the easiest solution by my opinion.